static/js/containers/CreatePostPage.js

```diff
@@ -81,26 +81,38 @@ class CreatePostPage extends React.Component<CreatePostPageProps> {
     ensureTwitterEmbedJS()
   }
 
+  shouldTriggerUpdate() {
+    const { postForm } = this.props
+    return (
+      postForm &&
+      emptyOrNil(postForm.value.text + postForm.value.url) &&
+      !emptyOrNil(postForm.value.postType)
+    )
+  }
+
   componentDidUpdate(prevProps: CreatePostPageProps) {
     const { channel, dispatch, postForm } = this.props
 
     // we may need to change out the postType under certain conditions
     // if the user switches channels and the new channel isn't LINK_TYPE_ANY,
     // and the current postType is null or a postType which is incompatible with the
-    // new channel, we want to set the postType to be the link_type for the channel
+    // new channel, we want to set the postType to be the link_type for the channel.
+    // If the new channel is LINK_TYPE_ANY and both text and url values are empty,
+    // we want to set the postType to null to display the post type options.
     if (
       postForm &&
       channel &&
       ((prevProps.channel && prevProps.channel.name !== channel.name) ||
         !prevProps.channel) &&
-      channel.link_type !== LINK_TYPE_ANY &&
+      (channel.link_type !== LINK_TYPE_ANY || this.shouldTriggerUpdate()) &&
```

Comment:
This boolean and the comment above it were confusing me enough that I tinkered with it in a branch. This is what I came up with:

```javascript
    // If a newly-selected channel cannot be rendered with the same form state as the
    // previously-selected channel, we need to reset the form state.
    if (
      postForm &&
      channel &&
      prevProps.channel &&
      // The selected channel's link type is different from the previously-selected channel
      channel.link_type !== prevProps.channel.link_type &&
      (
        // The selected channel accepts any type of post, and no input has been entered
        (channel.link_type === LINK_TYPE_ANY && R.all(emptyOrNil)(postForm.value.url, postForm.value.text)) ||
        // The selected channel only takes specific posts, and the post type of the
        // current form doesn't match the channel's post type
        (channel.link_type !== LINK_TYPE_ANY && channel.link_type !== postForm.value.postType)
      )
    ) {
```

It's still a big statement, but to me it's much more clear about what should trigger a form reset. As far as I could tell, the channel name comparison and the check for an empty post type are unnecessary. The 2 last conditionals (the ones that make up the 'or' expression) could be moved to helper functions, but I think it's OK leaving them inline with a comment. `R.all(emptyOrNil)` could be a new helper function if you feel like doing that. Let me know what you think about all that

Also, if you decide to go with something like this and the tests all explode in horrible ways, we could move this to a separate tech debt issue

Reply:
Using the above code, if you:
- Choose a text-only channel and enter some text
- Change to a channel that accepts both text and links

The text field will disappear along with your text.  This should not happen unless it is blank.  With the current code in the PR, it behaves as expected.

Reply:
Ah, the problem with with the `R.all` call. Should be:

`R.all(emptyOrNil, [postForm.value.url, postForm.value.text])`

Reply:
- Click 'Compose' from the home page
- Without selecting a channel, click 'Write something' to show the textarea but don't write anything.
- Now select a links-only channel.  The textarea does not disappear as it should.

Also a few unit tests break related to this scenario (from null postType to link/text).  I agree the current conditional logic is complex and not that easy to follow, but it seems to handle all the possible permutations correctly and passes all the tests, so I'd prefer to keep it as is.

Reply:
Sounds good. I'll create a tech debt issue. I appreciate your patience!

Reply:
No problem, thanks for taking the time to try finding a simpler solution.